_posts/2021-05-06-pixel-tests.md

```diff
@@ -0,0 +1,191 @@
+---
+title: Testing all the pixels
+author: mvo
+date: 2021-05-06 10:00
+tags: cockpit
+slug: pixel-testing
+category: tutorial
+comments: 'true'
+---
+
+The Cockpit integration tests can now contain "pixel tests". Such a
+ test will take a screenshot with the browser and compare it with a
+ reference.  The idea is that we can catch visual regressions much
+ easier this way than if we would hunt for them in a purely manual
+ fashion.
+
+## Preparing a repository for pixel tests
+
+A pixel test will take a screenshot of part of the Cockpit UI and
+compare it with a reference.  Thus, these reference images are
+important and play the biggest role.
+
+A large part of dealing with pixel tests will consequently consist of
+maintaining the reference images.  At the same time, we don't want to
+clog up our main source repository with them.  While the number and
+size of the reference images at any one point in time should not pose
+a problem, we will over time accumulate a history of them that we are
+afraid would dominate the source repository.
+
+Thus, the reference images are not stored in the source repository.
+Instead, we store them in an external repository that is linked into
+the source repository as a submodule.  That external repository
+doesn't keep any history and can be aggressively pruned.
+
+Developers are mostly isolated from this via the new
+`test/common/pixel-tests` tool.  But if you are familiar with git
+submodules, there should be no suprises for you here.
+
+A source repository needs to be prepared before it can store reference
+images in a external storage repository. You can let the tool do it by
+running
+
+```
+./test/common/pixel-tests init
+```
+
+and then committing the changes it has done to the source repository.
+(Those changes will be a new or modified .gitmodules file, and a new
+gitlink at test/reference.)
+
+## Adding a pixel test
+
+To add a pixel test to a test program, call the new "assert_pixels"
+function of the Browser class:
+
+     ...
+     self.browser.assert_pixels('#dialog button.apply', "button")
+     ...
+
+The first argument is a CSS selector that identifies the part of the
+UI that you want to compare.  The screenshot will only include that
+element.  The second argument is an arbitrary but unique key for this
+test point.  It is used to name the files that go with it.
+
+For each such call, there needs to be a reference image in the
+`test/reference/` directory.
+
+As mentioned above, the `test/reference/` directory is very special
+indeed, and needs to be carefully managed with the
+`test/common/pixel-tests` tool (or even more carefully with `git
+submodule` et al).
+
+First, make sure `test/reference` is up-to-date:
+
+```
+$ ./test/common/pixel-tests pull
```

Comment:
Drop the `$` for consistency with the other commands above.

Reply:
Hmm, but they all have a `$`.  Except the first, which I have now fixed.